Variable simple range

Jakobeha opened the pull request:
Addresses #789. Note that this may slow performance because we still usually have to compile the regular for, unless we can statically resolve that the lower bound will behave like an integer (which is hard because usually it's a real, and it would need to be rounded to an integer if so)

Jakobeha committed:
initial attempt in RIR

Jakobeha committed:
fixed stuff and implemented in PIR
one note: we reuse IsType for the integer-castable test. This makes sense in RIR but not much overlap in PIR (it completely subsumes typeFeedback), perhaps we should create a new instruction?

Jakobeha committed:
added test

Jakobeha committed:
update instructions
- add
- remove  and
- revert

Jakobeha committed:
simple for in Compiler.cpp

Jakobeha committed:
slight changes to interp.cpp

Jakobeha committed:
fix potential typefeedback bug
not sure if the fix is correct, it's probably because I refactored
getting typeChecks in a previous approach and didn't change it back

Jakobeha committed:
rewriting simple range...

Jakobeha committed:
fixed stack allocation

Jakobeha committed:
updated pir_check

minor type inference fix

o- commented on rir/src/ir/Compiler.cpp:
```diff
@@ -224,112 +224,117 @@ void emitGuardForNamePrimitive(CodeStream& cs, SEXP fun) {
+            cs << BC::colonCastRhs();
+
+            // step <- colonGetStep(m', n')
+            cs << BC::colonGetStep();
```
I think the old version of explicitly computing `step = m<n ? 1 : -1` was better, because pir can speculate on the branch (thanks to `recordTest`) and then constantfold step. this is not possible anymore here.

o- replied:
to be clear, I am proposing to remove the colonGetStep instruction in favor if just a branch on `lt` that pushes either `1L` or `-1L`

o- commented on rir/src/compiler/native/lower_llvm.cpp:
```diff
@@ -4725,6 +4687,14 @@ bool LowerFunctionLLVM::tryCompile() {
+            case Tag::ColonCastLhs:
+            case Tag::ColonCastRhs:
+            case Tag::ColonGetStep:
+                // TODO: Handle via native builtins
```
this needs to be implemented, or you disable the native backend for any function with a for loop, that's not good...

o- commented on rir/src/compiler/opt/cleanup.cpp:
```diff
@@ -130,13 +130,13 @@ class TheCleanup {
+                    auto rhs = colonInputEffects->arg<1>().val();
+                    if (!lhs->type.maybeHasAttrs() ||
+                        !rhs->type.maybeHasAttrs()) {
+                        colonInputEffects->replaceUsesWith(True::instance());
```
move to constantfolding.

o- commented on rir/src/ir/Compiler.cpp:
```diff
@@ -224,112 +224,117 @@ void emitGuardForNamePrimitive(CodeStream& cs, SEXP fun) {
+            cs << BC::force();
+            // if (!colonInputEffects(m, n)) {
+            cs << BC::colonInputEffects();
+            cs.addSrc(seq);
```
putting a `recordTest` here allows the compiler to speculate on dead branches more easily

o- commented on rir/src/compiler/pir/instruction.h:
```diff
@@ -1134,17 +1134,79 @@ class FLI(AsTest, 1, Effects() | Effect::Error | Effect::Warn) {
-    bool ceil;
+    explicit ColonInputEffects(Value* lhs, Value* rhs, unsigned srcIdx)
+        : FixedLenInstruction(NativeType::test,
+                              {{PirType::any(), PirType::any()}}, {{lhs, rhs}},
```
this instruction expects evaluated values, not promises. so the input types should be `PirType::val()` and not `any()`

o- replied:
(same for the other instrs)

o- commented on rir/src/compiler/util/type_test.h:
```diff
@@ -33,9 +33,10 @@ class TypeTest {
-        if (expected.noAttribs().isA(RType::integer) ||
-            expected.noAttribs().isA(RType::real) ||
-            expected.noAttribs().isA(RType::logical)) {
+        if (!expected.maybeObj() && // TODO: Is this right?
```
I think currently noAttrib implies noObj, but it's more robust, so I think it's ok.

o- reviewed (commented):
nice PR!
can you constantfold the castLhs/Rhs one too if the inputs are numerical? not sure, needs some thought.

Jakobeha committed:
review (and settings file change)
also added more constantfolding and fixed print_interp printing stack
for the native backend, we simply call the same code used in interp.cpp, no IR generation

Jakobeha committed:
fixed native backend
removed Bool type because it was confusing, maybe that should be kept or in a separate PR?
